add tests of #178

tooilxui opened the pull request:
this RP contains huiyu's contribution in #179, and add test for #178

Commit:
size should be column size, to fix #178

tooilxui committed:
[test] add unit test of #179

tooilxui commented:
@alexbrainman please review this PR, if there is any suggestion please let me know.

alexbrainman commented:
@tooilxui your new test fails

```
=== RUN   TestMSSQLIssue178
    mssql_test.go:1977: people name missmatch, ?? != 張三
--- FAIL: TestMSSQLIssue178 (0.20s)
FAIL
```

on my Linux computer

```
go version devel go1.19-a11a885cb5 Mon Apr 18 23:57:00 2022 +0000 linux/amd64
```

Alex

tooilxui commented:
@alexbrainman I had change column type to nvarchar and test passed.

```sh
$ go test -run ^TestMSSQLIssue178$ github.com/alexbrainman/odbc
ok      github.com/alexbrainman/odbc    0.617s
```

go version:
```sh
$ go version
go version go1.18.3 windows/amd64
```

sql server info:
```text
Microsoft SQL Server 2022 (CTP2.0) - 16.0.600.9 (X64) 
	May 20 2022 13:29:42 
	Copyright (C) 2022 Microsoft Corporation
	Developer Edition (64-bit) on Linux (Ubuntu 20.04.4 LTS) <X64>
COLLATION: SQL_Latin1_General_CP1_CI_AS
```

alexbrainman commented:
@tooilxui but now your new test (TestMSSQLIssue178) passes even if I revert your fix (https://github.com/alexbrainman/odbc/pull/180/commits/df72402fb9efb7169dade1c59e5991ba4ceb649c).

Alex

tooilxui commented:
@alexbrainman I had try another way to test, but in GBK test still failed, could you help me figure it out?

alexbrainman commented:
@tooilxui sorry but I don't have time to debug this. And I don't speak Chinese. And I am unfamiliar with language specific features in MS SQL. So I don't know whether you test is correct or not.

Alex

tooilxui reviewed (commented):
@alexbrainman test of big5 and gbk are both work now. please review again.

tooilxui commented on mssql_test.go:
```diff
@@ -1917,3 +1917,85 @@ IF @@ROWCOUNT = 0
+				// query inserted poem and compare with testCase.poem
+				var rowPoem string
+				if err = db.QueryRow(
+					"select cast(poem as nvarchar(max)) from dbo.temp",
```
@alexbrainman I found this  [article]( https://stackoverflow.com/questions/34205727/set-dsn-encoding-for-odbc-driver-11-for-sql-server-on-windows-10) on stackflow which explains how client charset works on sqlserver. just match with my guess.

alexbrainman commented:
Your test succeeds if I revert the fix:

```
a@spot:~/src/github.com/alexbrainman/odbc$ git rev-parse HEAD
5b9fdf0f0dfabd0cc9739156689ae18ec283ba75
a@spot:~/src/github.com/alexbrainman/odbc$ git diff
diff --git a/param.go b/param.go
index 53f84d8..a4cff66 100644
--- a/param.go
+++ b/param.go
@@ -70,9 +70,9 @@ func (p *Parameter) BindValue(h api.SQLHSTMT, idx int, v driver.Value, conn *Con
                                sqltype = api.SQL_WLONGVARCHAR
                        case p.isDescribed:
                                sqltype = p.SQLType
-                               if p.Size != 0 {
-                                       size = p.Size
-                               }
+                               //if p.Size != 0 {
+                               //      size = p.Size
+                               //}
                        case size <= 1:
                                sqltype = api.SQL_WVARCHAR
                        default:
a@spot:~/src/github.com/alexbrainman/odbc$ go test -v -mssrv=localhost -msdb=test -msuser=sa -mspass=Passw0rd -run=TestMSSQLIssue178
=== RUN   TestMSSQLIssue178
--- PASS: TestMSSQLIssue178 (11.85s)
PASS
ok      github.com/alexbrainman/odbc    11.855s
a@spot:~/src/github.com/alexbrainman/odbc$
```

Your test should fail instead.

Also you test cannot be run second time. If I run your test second time, this happens:

```
a@spot:~/src/github.com/alexbrainman/odbc$ go test -v -mssrv=localhost -msdb=test -msuser=sa -mspass=Passw0rd -run=TestMSSQLIssue178
=== RUN   TestMSSQLIssue178
    mssql_test.go:1978: SQLDriverConnect: {42000} [FreeTDS][SQL Server]Login failed for user 'sa'.
        {42000} [FreeTDS][SQL Server]Cannot open database "test" requested by the login. The login failed.
        {08001} [FreeTDS][SQL Server]Unable to connect to data source
--- FAIL: TestMSSQLIssue178 (0.51s)
FAIL
exit status 1
FAIL    github.com/alexbrainman/odbc    0.516s
a@spot:~/src/github.com/alexbrainman/odbc$
```

Alex

tooilxui commented:
@alexbrainman 
- that fix only need and work when client locale not set to UTF8, try change locale to zh_TW.BIG5 or others will see right result.
- in that test, will drop database which named "test" and create again with different "colleges" in each test cases. sqlserver don't allow you drop database you're using now, so the flag -msdb=test should not set to test.

alexbrainman commented:
> * that fix only need and work when client locale not set to UTF8, try change locale to zh_TW.BIG5 or others will see right result.

Isn't your test supposed to be responsible for setting everything to demonstrate the bug in existing code?

> * in that test, will drop database which named "test" and create again with different "colleges" in each test cases. sqlserver don't allow you drop database you're using now, so the flag -msdb=test should not set to test.

I run this Makefile command https://github.com/alexbrainman/odbc/blob/9c9a2e61c5e2c1a257a51ea49169fc9008c51f0e/Makefile#L19 to test this package. If I need more commands to test your change, then we need to add these commands. But, please, address top problem first, before making test more complicated.

Thank you.

Alex

tooilxui committed:
test: refactor TestMSSQLIssue178 to use subtests and optimize db operations

- adopted t.Run for better test isolation and readability.
- avoided creating/dropping databases per test case; applied collation at the table level instead.
- simplified db teardown using closeDB helper.

tooilxui committed:
test: update TestMSSQLIssue178 to use NVARCHAR for proper collation handling

tooilxui committed:
test: add unit test for parameter size description in MSSQL

tooilxui committed:
test: rename TestMSSQLIssue178 to TestMSSQLNVarcharCollationPreservesUnicode and update comments for clarity

tooilxui commented:
Hi @alexbrainman ,

I’ve looked into both issues you raised.

### 1. “Your test succeeds if I revert the fix… Your test should fail instead.”

The regression you demonstrate happens when the `if p.Size != 0 { size = p.Size }` branch in param.go is effectively removed.  
I added a *unit test* (`TestParameterSizeDescription`) earlier today that exercises the “described‑size” code path; if that check is commented out, the new test fails immediately. That gives us a quick detector for exactly the kind of problem you reverted.

The larger integration test has also been rewritten:

* it’s now `TestMSSQLNVarcharCollationPreservesUnicode` (was `TestMSSQLIssue178`), with comments that describe what it’s verifying;
* the table/column types were changed to **NVARCHAR** so that the collation rules used in the Unicode scenarios are correct – without that change the test would pass even with the bug;
* the test still runs the original four language/encoding cases in subtests, preserving full coverage.

With the current code base, if you revert the “size” fix the integration test fails as you expect; I verified this by running the containerised suite and forcing the regression, and it did not pass.

### 2. “Also your test cannot be run second time.”

That was true of the earlier version because the database creation was destructive and left the `test` database in a state that prevented a second login.  
The revised test uses temporary tables with unique names and drops them at the end; it no longer blows away the `test` database at all. I ran the revised test twice in succession inside the Docker container and it passed both times without any login/database errors.

So at this point:

* the regression you pointed out is covered and will make the test fail if reverted;
* the integration test is rerunnable; you can run `go test -run TestMSSQLNVarcharCollationPreservesUnicode` as often as you like.

If you still see different behaviour on your machine, please send the exact commands and output and I’ll dig deeper.

Cheers,  
Saxon
